### PR TITLE
[DataGrid] Warn when pageSize is not present in rowsPerPageOptions

### PR DESCRIPTION
**Pull request by flaviendelangle**
One chunk of #1907

### CODE REVIEW

**Comment by flaviendelangle** [comment]
The snapshots changes are expected, we don't pass `rawPerPageOptions` as props so the default one are used (`DEFAULT_GRID_OPTIONS.rowsPerPageOptions`)

**Comment by flaviendelangle** [comment]
@oliviertassinari , I updated the behavior as discussed yesterday :+1:

**Comment by flaviendelangle** [comment]
Their is an error on a test because the effect runs with the default `pageSize` and then with the value passed by the user.
It should be fixed by the clean controlled state. It put this PR on hold until it's done :+1:

**Review by oliviertassinari — COMMENTED** [review]
👍 for the current direction

File: packages/grid/_modules_/grid/components/GridPagination.tsx
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridPagination.tsx:91]: You're right, and running it in the render would also avoid this issue https://github.com/mui-org/material-ui-x/pull/2014#issuecomment-874717496

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridPagination.tsx]: My bad, it's not enough
If I put it in the render, on the 1st render I have the default pagination.
So I'll have to wait to the controlled mode rework.

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:327]: Just a small thing. We stopped using `React.FC` in #1325.

```suggestion
        const TestCase = () => {
```

Comment (flaviendelangle) [reply packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx]: We can use `React.VoidFunctionComponent` to exclude the children

**Comment by oliviertassinari** [comment]
@flaviendelangle I have rebased and updated the tests to pass. A codesandbox with the new DX: https://codesandbox.io/s/material-demo-forked-pjlyc?file=/demo.tsx. However, having to include 100 in the list, that's not flying 😁. The simplest might be to add a hack for 100 and to leave a TODO.

Updating the options in an effect? It's borderline. I think that it would only work under two conditions: 1. We initiate the state with the right initial options (KO currently) 2. We make the child of the state provider a memo, to always stop the React rendering so that children only render once per option updates (KO currently).

Otherwise, we can use React, create a dedicated context for the options (top-level props).

**Comment by m4theushw** [comment]
In #1720 a new `GridRootPropsContext` context was added with the root props. You can use it instead of querying the state. We get for free the advantage of not having to add 100 to the list. You just have to solve the dependency cycle, maybe add the prop you need in the parent element and pass it to the pagination component.

**Comment by flaviendelangle** [comment]
I'll finish this one once the controlled pagination is merged
And I agree, looking at the input props should help

**Comment by flaviendelangle** [comment]
https://app.circleci.com/pipelines/github/mui-org/material-ui-x/6357/workflows/3eb1a423-62eb-40b7-b3ff-5cf21a9beef9/jobs/33830

@oliviertassinari if I set a global variable to only warn once, it's not reset between two tests so 2 of the 3 tests don't have the warning.
I'm using a ref to bypass this issue.

File: packages/grid/_modules_/grid/hooks/features/useGridComponents.tsx
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/useGridComponents.tsx:55]: Same reason as https://github.com/mui-org/material-ui-x/pull/2213/files#r676538144
It's to avoid circle import

**Comment by oliviertassinari** [comment]
@flaviendelangle 

Capture d’écran 2021-07-27 à 12 15 29

You have discarded one of my commit. Is it on purpose?

File: packages/grid/_modules_/grid/hooks/features/useGridComponents.tsx
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/useGridComponents.tsx:3]:
```suggestion
```

**Review by oliviertassinari — COMMENTED** [review]
Looks good

**Comment by flaviendelangle** [comment]
@oliviertassinari what was the goal of adding the `100` in `rowsPerPageOptions` ?

**Comment by oliviertassinari** [comment]
@flaviendelangle To dodge the initial value issue. You fixed this using the root props context.

File: packages/grid/_modules_/grid/components/GridPagination.tsx
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/GridPagination.tsx:91]: We can give a default value for the prop in:

https://github.com/mui-org/material-ui-x/blob/75ae2f73204939c73ae3b8f5597b6541fefe34dc/packages/grid/data-grid/src/DataGrid.tsx#L53

The same for the XGrid.

```suggestion
      !rootProps.rowsPerPageOptions!.includes(rootProps.pageSize)
```

File: packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx:365]: I wouldn't mind hardcoding the value if we want to avoid exporting constants.

```suggestion
        `Material-UI: The page size \`100\` is not preset in the \`rowsPerPageOptions\``,
```

File: packages/grid/_modules_/grid/components/GridPagination.tsx
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/GridPagination.tsx:91]: If we do so, then `useGridPageSize` will always be in controlled mode no ?
We could give the default props only to the context and not to the feature hooks but I don't think it's a good idea.

Comment (m4theushw) [reply packages/grid/_modules_/grid/components/GridPagination.tsx:91]: We can use the state value, If think. One test related to auto size will fail, maybe we don't display the warning in this condition because there's no way to change the rows per page.

```suggestion
      !rootProps.rowsPerPageOptions!.includes(rootProps.pageSize ?? paginationState.pageSize)
```

**Review by oliviertassinari — COMMENTED** [review]
It looks clean

**Review by m4theushw — APPROVED** [review]

**Pull request merged**
